File: app01_java_block/src/test/java/com/gandarych/legalchain/p2p/TwoNodeSyncIntegrationTest.java
package com.gandarych.legalchain.p2p;

import com.gandarych.legalchain.LegalChainApplication;
import com.gandarych.legalchain.core.Blockchain;
import com.gandarych.legalchain.core.NodeWallet;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.context.ConfigurableApplicationContext;

import java.util.function.BooleanSupplier;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

/**
 * Automated version of the two-node demo from the Definition of Done: two full
 * Spring Boot nodes in one JVM, connected over a real WebSocket with the real
 * ML-DSA + ML-KEM handshake, converging on one ledger.
 *
 * <p>What this proves end-to-end: distinct self-sovereign node identities (BE-14),
 * a deterministic shared genesis (BE-02), a PQC-secured channel established without
 * any pre-shared secret (SEC-04), and full-chain adoption where the receiving node
 * re-audits every block, signature and proof locally before accepting (E4).</p>
 */
@SpringBootTest(
        webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
        properties = "legalchain.node.name=sync-itest-A")
class TwoNodeSyncIntegrationTest {

    @LocalServerPort
    private int portA;

    @Autowired
    private Blockchain chainA;
    @Autowired
    private NodeWallet walletA;
    @Autowired
    private P2pService p2pA;

    @Test
    void twoRemoteNodesEstablishPqcChannelAndConvergeOnOneLedger() {
        ConfigurableApplicationContext nodeB = new SpringApplicationBuilder(LegalChainApplication.class)
                .properties("server.port=0", "legalchain.node.name=sync-itest-B")
                .run();
        try {
            Blockchain chainB = nodeB.getBean(Blockchain.class);
            NodeWallet walletB = nodeB.getBean(NodeWallet.class);

            // BE-14: each node generates its own ML-DSA identity — fingerprints differ.
            assertNotEquals(walletA.address(), walletB.address());
            // BE-02: yet both derive the identical deterministic genesis block.
            assertEquals(chainA.blocks().getFirst().hash(), chainB.blocks().getFirst().hash());

            // Grow node A's ledger past node B's.
            chainA.minePendingTransactions(walletA.address());
            chainA.minePendingTransactions(walletA.address());
            int targetLength = chainA.length();

            // B dials A: WebSocket + ML-DSA-signed hello + ML-KEM encapsulation.
            nodeB.getBean(P2pClientService.class).connect("ws://localhost:" + portA + "/ws/p2p");
            P2pService p2pB = nodeB.getBean(P2pService.class);
            await("PQC handshake on both sides",
                    () -> !p2pB.connectedPeers().isEmpty() && !p2pA.connectedPeers().isEmpty());

            // Request the ledger over the encrypted channel; B re-audits before adopting.
            p2pB.syncWithPeers();
            await("chain convergence",
                    () -> chainB.length() >= targetLength
                            && chainB.blocks().getLast().hash().equals(chainA.blocks().getLast().hash()));

            assertTrue(chainB.isValid(), "adopted chain passes B's own full audit");
            assertEquals(chainA.blocks().getLast().hash(), chainB.blocks().getLast().hash());
        } finally {
            nodeB.close();
        }
    }

    /** Polls the condition for up to 20 s — handshake and sync are asynchronous. */
    private static void await(String what, BooleanSupplier condition) {
        long deadline = System.currentTimeMillis() + 20_000;
        while (System.currentTimeMillis() < deadline) {
            if (condition.getAsBoolean()) {
                return;
            }
            try {
                Thread.sleep(200);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                fail("interrupted while waiting for " + what);
            }
        }
        fail("timed out after 20s waiting for " + what);
    }
}

File: app03_csharp_block/tests/LegalChain.Tests/TwoNodeSyncIntegrationTest.cs
using LegalChain.Core;
using LegalChain.P2p;
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.DependencyInjection;

namespace LegalChain.Tests;

/// <summary>
/// Automated version of the two-node demo from the Definition of Done: two full
/// ASP.NET Core nodes (real Kestrel, real WebSockets) in one process, connected with the
/// real ML-DSA + ML-KEM handshake, converging on one ledger.
/// </summary>
/// <remarks>
/// What this proves end-to-end: distinct self-sovereign node identities (BE-14),
/// a deterministic shared genesis (BE-02), a PQC-secured channel established without
/// any pre-shared secret (SEC-04), and full-chain adoption where the receiving node
/// re-audits every block, signature and proof locally before accepting (E4).
/// </remarks>
public class TwoNodeSyncIntegrationTest
{
    [Fact]
    public async Task TwoRemoteNodesEstablishPqcChannelAndConvergeOnOneLedger()
    {
        WebApplication nodeA = LegalChainApp.Build(
            ["--urls", "http://127.0.0.1:0", "--node-name", "sync-itest-A"]);
        WebApplication nodeB = LegalChainApp.Build(
            ["--urls", "http://127.0.0.1:0", "--node-name", "sync-itest-B"]);
        await nodeA.StartAsync();
        await nodeB.StartAsync();
        try
        {
            string urlA = nodeA.Urls.Single(); // ephemeral port resolved after StartAsync
            var chainA = nodeA.Services.GetRequiredService<Blockchain>();
            var chainB = nodeB.Services.GetRequiredService<Blockchain>();
            var walletA = nodeA.Services.GetRequiredService<NodeWallet>();
            var walletB = nodeB.Services.GetRequiredService<NodeWallet>();
            var p2pA = nodeA.Services.GetRequiredService<P2pService>();
            var p2pB = nodeB.Services.GetRequiredService<P2pService>();

            // BE-14: each node generates its own ML-DSA identity — fingerprints differ.
            Assert.NotEqual(walletA.Address, walletB.Address);
            // BE-02: yet both derive the identical deterministic genesis block.
            Assert.Equal(chainA.Blocks()[0].Hash, chainB.Blocks()[0].Hash);

            // Grow node A's ledger past node B's.
            chainA.MinePendingTransactions(walletA.Address);
            chainA.MinePendingTransactions(walletA.Address);
            int targetLength = chainA.Length;

            // B dials A: WebSocket + ML-DSA-signed hello + ML-KEM encapsulation.
            string wsUrl = urlA.Replace("http://", "ws://") + "/ws/p2p";
            await nodeB.Services.GetRequiredService<P2pClientService>().ConnectAsync(wsUrl);
            await AwaitAsync("PQC handshake on both sides",
                () => p2pB.ConnectedPeers().Count > 0 && p2pA.ConnectedPeers().Count > 0);

            // Request the ledger over the encrypted channel; B re-audits before adopting.
            await p2pB.SyncWithPeersAsync();
            await AwaitAsync("chain convergence",
                () => chainB.Length >= targetLength
                      && chainB.Blocks()[^1].Hash == chainA.Blocks()[^1].Hash);

            Assert.True(chainB.IsValid(), "adopted chain passes B's own full audit");
            Assert.Equal(chainA.Blocks()[^1].Hash, chainB.Blocks()[^1].Hash);
        }
        finally
        {
            await nodeA.StopAsync();
            await nodeB.StopAsync();
        }
    }

    /// <summary>Polls the condition for up to 20 s — handshake and sync are asynchronous.</summary>
    private static async Task AwaitAsync(string what, Func<bool> condition)
    {
        var deadline = DateTime.UtcNow.AddSeconds(20);
        while (DateTime.UtcNow < deadline)
        {
            if (condition())
            {
                return;
            }
            await Task.Delay(200);
        }
        Assert.Fail("timed out after 20s waiting for " + what);
    }
}
